### DISCUSSION TITLE
Java-client v8.x - How to add listeners to an instance of `AppiumDriver`?

### DISCUSSION THREAD

**Discussion by cliffgibb** [appium/java-client#1760] [Q&A]
I am in the process of migrating from java client v7.x to v8.x and need to update the process of attaching listeners to the instance of `AppiumDriver`.

**Old v7.x method:**
`AppiumDriver driver;` 
(initialization omitted)
`driver = EventFiringWebDriverFactory.getEventFiringWebDriver(driver, new AppiumDriverEventListener());`

**New v8.x method:**
`AppiumDriver driver;` 
(initialization omitted)
```
WebDriverListener listener = new AppiumDriverEventListener();
driver = (AppiumDriver) new EventFiringDecorator(listener).decorate(driver);
```

I am not able to cast the `WebDriver` to `AppiumDriver`:
java.lang.ClassCastException: class net.bytebuddy.renamed.java.lang.Object$ByteBuddy$4VFg5tn3 cannot be cast to class io.appium.java_client.AppiumDriver (net.bytebuddy.renamed.java.lang.Object$ByteBuddy$4VFg5tn3 is in unnamed module of loader net.bytebuddy.dynamic.loading.ByteArrayClassLoader @27d98bef; io.appium.java_client.AppiumDriver is in unnamed module of loader com.miw.mcb.shared.classloaders.JarClassLoader @1992eaf4)

What is the correct approach to add listeners to an instance of `AppiumDriver`?

**Comment by mykola-mokhnach** [comment]
`driver` must be an interface in the code above. Assigning it to a class instance won't work

**Comment by mykola-mokhnach** [comment]
> driver = (AppiumDriver) new EventFiringDecorator(listener).decorate(driver);

also this typecast won't work, the result of the `decorate` method could only be cast to an interface

**Reply by cliffgibb** [reply]
Thank you for your response @mykola-mokhnach. Since my implementation relies on an instance of AppiumDriver, do you have a recommendation as how I can add listeners to an instance of AppiumDriver?

I could do this using v7.x:
`driver = EventFiringWebDriverFactory.getEventFiringWebDriver(driver, new AppiumDriverEventListener());`

I can't simply replace AppiumDriver with WebDriver.

Would this be considered a bug of v8.x? (I'd be happy to submit a ticket) or perhaps there is another approach of which I am not aware.

**Reply by mykola-mokhnach** [reply]
No, the behaviour is expected.
You could create an improvement request to https://github.com/SeleniumHQ/selenium though as we import the decorator implementation from there without any changes.

**Reply by cliffgibb** [reply]
Hi @mykola-mokhnach, it appears that Selenium has made a fix in version 4.5.0 (the `EventFiringDecorator` accepts a target Class):
https://github.com/SeleniumHQ/selenium/commit/fcc549ad290d9ae13beecfd875b57178d5fcb2c5

So once the Appium java-client uses the Selenium 4.5.0 dependency, I will be able to give this a test.
